core-tests/resource-release-test.sh

```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+
+set -ux
+
+# This assumes that the compiled resource-release-test executable
+# is in the PATH
+
+# create a folder for the duration of the test, and make sure it starts empty
+DIR=`dirname $0`/resource-release-test-files
+rm -rf "$DIR"
+mkdir "$DIR"
+function finish {
+  rm -rf "$DIR"
+}
+trap finish EXIT
+
+# start resource-release-test and wait until it's ready
+resource-release-test "$DIR" &
+PID="$!"
+while [ ! -f "$DIR/test-has-started" ]; do sleep 1; done
+
+# kill resource-release-test and wait for it to release its resources.
+kill -s INT "$PID"
```

Comment:
I am trying to reproduce the "hit Ctrl-C while the test is sleeping" behaviour by sending a SIGINT signal when I know, via other means, that this is where the test's execution is stuck. Tracing indicates that this successfully causes the `finally` block to release its resources, but for reasons I don't understand, `defaultMain` doesn't terminate once the exception finishes propagating up the stack. So this test never terminates.

Reply:
I now understand that `defaultMain` wasn't terminating because the worker threads were not terminating, see my top-level comment.